File: frontend/src/components/benchmark/content/dfs/implementations/dfs.java
import java.util.*;

public class DFS {
    private Map<Integer, Set<Integer>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    public void addEdge(int u, int v) {
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.putIfAbsent(v, new HashSet<>());
        adjacency.get(u).add(v);
        adjacency.get(v).add(u);
        nodes.add(u);
        nodes.add(v);
    }

    public void addDirectedEdge(int u, int v) {
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.putIfAbsent(v, new HashSet<>());
        adjacency.get(u).add(v);
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edges, boolean directed) {
        for (int[] edge : edges) {
            if (directed) {
                addDirectedEdge(edge[0], edge[1]);
            } else {
                addEdge(edge[0], edge[1]);
            }
        }
    }

    private void dfsRecursive(int node, Set<Integer> visited, List<Integer> order) {
        visited.add(node);
        order.add(node);

        if (!adjacency.containsKey(node)) return;

        for (int neighbor : adjacency.get(node)) {
            if (!visited.contains(neighbor)) {
                dfsRecursive(neighbor, visited, order);
            }
        }
    }

    public List<Integer> dfsRecursive(int start) {
        List<Integer> order = new ArrayList<>();
        Set<Integer> visited = new HashSet<>();
        dfsRecursive(start, visited, order);
        return order;
    }

    public List<Integer> dfsIterative(int start) {
        List<Integer> order = new ArrayList<>();
        Set<Integer> visited = new HashSet<>();
        Stack<Integer> stack = new Stack<>();

        stack.push(start);

        while (!stack.isEmpty()) {
            int node = stack.pop();

            if (visited.contains(node)) continue;

            visited.add(node);
            order.add(node);

            if (!adjacency.containsKey(node)) continue;

            for (int neighbor : adjacency.get(node)) {
                if (!visited.contains(neighbor)) {
                    stack.push(neighbor);
                }
            }
        }

        return order;
    }

    public List<Integer> dfsPath(int start, int goal) {
        Map<Integer, Integer> parent = new HashMap<>();
        Set<Integer> visited = new HashSet<>();
        Stack<Integer> stack = new Stack<>();

        parent.put(start, -1);
        stack.push(start);

        while (!stack.isEmpty()) {
            int node = stack.pop();

            if (visited.contains(node)) continue;
            visited.add(node);

            if (node == goal) {
                List<Integer> path = new ArrayList<>();
                while (node != -1) {
                    path.add(node);
                    node = parent.get(node);
                }
                Collections.reverse(path);
                return path;
            }

            if (!adjacency.containsKey(node)) continue;

            for (int neighbor : adjacency.get(node)) {
                if (!visited.contains(neighbor)) {
                    parent.put(neighbor, node);
                    stack.push(neighbor);
                }
            }
        }

        return new ArrayList<>();
    }

    public List<List<Integer>> findComponents() {
        List<List<Integer>> components = new ArrayList<>();
        Set<Integer> visited = new HashSet<>();

        for (int node : nodes) {
            if (!visited.contains(node)) {
                List<Integer> component = new ArrayList<>();
                dfsRecursive(node, visited, component);
                components.add(component);
            }
        }

        return components;
    }

    public boolean isConnected() {
        if (nodes.isEmpty()) return true;

        int start = nodes.iterator().next();
        Set<Integer> visited = new HashSet<>();
        List<Integer> order = new ArrayList<>();
        dfsRecursive(start, visited, order);

        return visited.size() == nodes.size();
    }

    public boolean hasCycle() {
        Set<Integer> visited = new HashSet<>();
        Set<Integer> recStack = new HashSet<>();

        for (int node : nodes) {
            if (!visited.contains(node)) {
                if (hasCycleUtil(node, visited, recStack, -1)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean hasCycleUtil(int node, Set<Integer> visited, Set<Integer> recStack, int parent) {
        visited.add(node);
        recStack.add(node);

        if (!adjacency.containsKey(node)) return false;

        for (int neighbor : adjacency.get(node)) {
            if (recStack.contains(neighbor) && neighbor != parent) {
                return true;
            }

            if (!visited.contains(neighbor)) {
                if (hasCycleUtil(neighbor, visited, recStack, node)) {
                    return true;
                }
            }
        }

        recStack.remove(node);
        return false;
    }
}

File: frontend/src/components/benchmark/content/dfs/implementations/dfs.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class DFS
{
    private Dictionary<int, HashSet<int>> adjacency = new Dictionary<int, HashSet<int>>();
    private HashSet<int> nodes = new HashSet<int>();

    public void AddEdge(int u, int v)
    {
        if (!adjacency.ContainsKey(u))
            adjacency[u] = new HashSet<int>();
        if (!adjacency.ContainsKey(v))
            adjacency[v] = new HashSet<int>();

        adjacency[u].Add(v);
        adjacency[v].Add(u);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void AddDirectedEdge(int u, int v)
    {
        if (!adjacency.ContainsKey(u))
            adjacency[u] = new HashSet<int>();
        if (!adjacency.ContainsKey(v))
            adjacency[v] = new HashSet<int>();

        adjacency[u].Add(v);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void BuildFromEdges(List<(int, int)> edges, bool directed = false)
    {
        foreach (var (u, v) in edges)
        {
            if (directed)
                AddDirectedEdge(u, v);
            else
                AddEdge(u, v);
        }
    }

    private void DfsRecursive(int node, HashSet<int> visited, List<int> order)
    {
        visited.Add(node);
        order.Add(node);

        if (!adjacency.ContainsKey(node)) return;

        foreach (int neighbor in adjacency[node])
        {
            if (!visited.Contains(neighbor))
            {
                DfsRecursive(neighbor, visited, order);
            }
        }
    }

    public List<int> DfsRecursive(int start)
    {
        var order = new List<int>();
        var visited = new HashSet<int>();
        DfsRecursive(start, visited, order);
        return order;
    }

    public List<int> DfsIterative(int start)
    {
        var order = new List<int>();
        var visited = new HashSet<int>();
        var stack = new Stack<int>();

        stack.Push(start);

        while (stack.Count > 0)
        {
            int node = stack.Pop();

            if (visited.Contains(node)) continue;

            visited.Add(node);
            order.Add(node);

            if (!adjacency.ContainsKey(node)) continue;

            foreach (int neighbor in adjacency[node])
            {
                if (!visited.Contains(neighbor))
                {
                    stack.Push(neighbor);
                }
            }
        }

        return order;
    }

    public List<int> DfsPath(int start, int goal)
    {
        var parent = new Dictionary<int, int>();
        var visited = new HashSet<int>();
        var stack = new Stack<int>();

        parent[start] = -1;
        stack.Push(start);

        while (stack.Count > 0)
        {
            int node = stack.Pop();

            if (visited.Contains(node)) continue;
            visited.Add(node);

            if (node == goal)
            {
                var path = new List<int>();
                while (node != -1)
                {
                    path.Add(node);
                    node = parent[node];
                }
                path.Reverse();
                return path;
            }

            if (!adjacency.ContainsKey(node)) continue;

            foreach (int neighbor in adjacency[node])
            {
                if (!visited.Contains(neighbor))
                {
                    parent[neighbor] = node;
                    stack.Push(neighbor);
                }
            }
        }

        return new List<int>();
    }

    public List<List<int>> FindComponents()
    {
        var components = new List<List<int>>();
        var visited = new HashSet<int>();

        foreach (int node in nodes)
        {
            if (!visited.Contains(node))
            {
                var component = new List<int>();
                DfsRecursive(node, visited, component);
                components.Add(component);
            }
        }

        return components;
    }

    public bool IsConnected()
    {
        if (nodes.Count == 0) return true;

        int start = nodes.First();
        var visited = new HashSet<int>();
        var order = new List<int>();
        DfsRecursive(start, visited, order);

        return visited.Count == nodes.Count;
    }

    public bool HasCycle()
    {
        var visited = new HashSet<int>();
        var recStack = new HashSet<int>();

        foreach (int node in nodes)
        {
            if (!visited.Contains(node))
            {
                if (HasCycleUtil(node, visited, recStack, -1))
                    return true;
            }
        }
        return false;
    }

    private bool HasCycleUtil(int node, HashSet<int> visited, HashSet<int> recStack, int parent)
    {
        visited.Add(node);
        recStack.Add(node);

        if (!adjacency.ContainsKey(node)) return false;

        foreach (int neighbor in adjacency[node])
        {
            if (recStack.Contains(neighbor) && neighbor != parent)
                return true;

            if (!visited.Contains(neighbor))
            {
                if (HasCycleUtil(neighbor, visited, recStack, node))
                    return true;
            }
        }

        recStack.Remove(node);
        return false;
    }
}